org.eclipse.jdt.core.manipulation/refactoring/org/eclipse/jdt/internal/corext/refactoring/sef/SelfEncapsulateFieldCompositeRefactoring.java

```diff
@@ -0,0 +1,229 @@
+		}
+		return fInitialConditions;
+	}
+
+	@Override
+	public Change createChange(IProgressMonitor pm) throws CoreException, OperationCanceledException {
+		pm.beginTask(NO_NAME, changes.size());
+		pm.setTaskName(RefactoringCoreMessages.SelfEncapsulateField_create_changes);
+		final DynamicValidationRefactoringChange result= new DynamicValidationRefactoringChange(getDescriptor(), getName());
+		for (TextChange change : changes) {
+			result.add(change);
+			pm.worked(1);
+		}
+		pm.done();
+		return result;
+	}
+
+	public EncapsulateFieldDescriptor getDescriptor () {
+		final Map<String, String> arguments= new HashMap<>();
+		String project= null;
+		IJavaProject javaProject= getSelectedRefactorings().get(0).getField().getJavaProject();
+		if (javaProject != null)
+			project= javaProject.getElementName();
+		final String description= Messages.format(RefactoringCoreMessages.SelfEncapsulateField_descriptor_description_short, BasicElementLabels.getJavaElementName(getSelectedRefactorings().stream().map(refactoring -> refactoring.getField().getElementName()).collect(Collectors.joining(", ")))); //$NON-NLS-1$
+		int flags= JavaRefactoringDescriptor.JAR_MIGRATION | JavaRefactoringDescriptor.JAR_REFACTORING | RefactoringDescriptor.STRUCTURAL_CHANGE | RefactoringDescriptor.MULTI_CHANGE;
+		try {
+			boolean isAnonymousOrLocal = true;
+			for(SelfEncapsulateFieldRefactoring refactoring : getSelectedRefactorings()) {
+				IType declaring = refactoring.getField().getDeclaringType();
+				isAnonymousOrLocal &= declaring.isAnonymous() || declaring.isLocal();
+			}
+			if(isAnonymousOrLocal) {
+				flags|= JavaRefactoringDescriptor.JAR_SOURCE_ATTACHMENT;
+			}
+		} catch (JavaModelException exception) {
+			JavaManipulationPlugin.log(exception);
+		}
+		String comment = getSelectedRefactorings().stream().map(refactoring -> refactoring.getCommentsForDescriptor().asString()).collect(Collectors.joining(", ")); //$NON-NLS-1$
+		getSelectedRefactorings().forEach(refactoring -> refactoring.getArgumentsForDescriptor().forEach((key, value) -> arguments.merge(key, value, (v1, v2) -> v1 + ", " + v2))); //$NON-NLS-1$
+		final EncapsulateFieldDescriptor descriptor= RefactoringSignatureDescriptorFactory.createEncapsulateFieldDescriptor(project, description, comment, arguments, flags);
+
+		return descriptor;
+	}
+
+	@Override
+	public RefactoringStatus checkFinalConditions(IProgressMonitor pm) throws CoreException, OperationCanceledException {
+		clearResources();
+		RefactoringStatus fFinalConditions = new RefactoringStatus();
+		for(SelfEncapsulateFieldRefactoring selfEncapsulateFieldRefactoring : getSelectedRefactorings()) {
+			fFinalConditions.merge(selfEncapsulateFieldRefactoring.checkFinalConditions(pm));
+		}
+		makeDeclarationsPrivate();
+		TextChangeManager fChangeManager = new TextChangeManager();
+		for(Entry<ICompilationUnit, ASTRewrite> entry : fRewritersMap.entrySet()) {
+			TextChange change= fChangeManager.get(entry.getKey());
```

Comment:
Can fChangeManager ever find an existing entry?  If so, what happens to the MultiTextEdit that was already created for it?  If there can't be an existing entry, simply use the TextChangeManager.get() logic when an entry is not found.

Reply:
Yes there can be an existing entry. The fChangeManager.get method checks for if the key doesnt exist then it creates a new entry and returns it. And this entry can be of a compilation unit where all the fields are defined or some other compilation unit where they are referenced. We want to work on the same AST for each CU found. So it makes more sense to keep a local object which is specific to a set of change made by the refactoring. Hence, we should not make any changes here.